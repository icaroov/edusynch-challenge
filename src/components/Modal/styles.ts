import styled, { css } from 'styled-components'

export const Container = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;

    position: relative;

    color: ${theme.colors.white};

    > svg {
      position: absolute;
      top: 0;
      right: 0;

      width: 3rem;
      border: none;
      font-size: 0;
      transition: 0.3s;

      color: ${theme.colors.white};
      background-color: transparent;

      cursor: pointer;

      &:focus {
        outline-width: 1px;
        outline-style: dashed;
        outline-color: ${theme.colors.white};
      }

      &:hover {
        opacity: 0.7;
      }
    }

    @media screen and ${theme.breakpoints.tabletMax} {
      > svg {
        top: -20px;
      }
    }
  `}
`

export const Content = styled.section`
  ${({ theme }) => css`
    white-space: normal;

    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.normal};
    line-height: 1.8em;
  `}
`
