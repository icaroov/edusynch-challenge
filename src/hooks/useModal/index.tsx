import { createContext, useContext, useState } from 'react'

type ModalContextData = {
  handleOpenModal: () => void
  handleCloseModal: () => void
  open: boolean
}

type ModalContext = {
  children: React.ReactNode
}

const ModalContext = createContext<ModalContextData>({} as ModalContextData)

export const ModalProvider = ({ children }: ModalContext) => {
  const [open, setOpen] = useState(false)

  function handleOpenModal() {
    setOpen(true)
  }

  function handleCloseModal() {
    setOpen(false)
  }

  return (
    <ModalContext.Provider value={{ handleCloseModal, handleOpenModal, open }}>
      {children}
    </ModalContext.Provider>
  )
}

export function useModal() {
  const context = useContext(ModalContext)

  if (!context) {
    throw new Error('useModal must be used within a ModalContextProvider')
  }

  return context
}
