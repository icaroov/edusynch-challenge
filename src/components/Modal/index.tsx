import ReactModal from 'react-modal'
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close'

import { useModal } from 'hooks/useModal'

import * as Styled from './styles'

const Modal = ({ children }: { children: JSX.Element }) => {
  const { open, handleCloseModal } = useModal()

  return (
    <ReactModal
      isOpen={open}
      onRequestClose={handleCloseModal}
      overlayClassName="react-modal-overlay"
      className="react-modal-content"
      contentLabel="Modal"
      ariaHideApp={false}
    >
      <Styled.Container>
        <CloseIcon aria-label="Close Modal" onClick={handleCloseModal} />

        <Styled.Content data-testid="content-modal">{children}</Styled.Content>
      </Styled.Container>
    </ReactModal>
  )
}

export default Modal
