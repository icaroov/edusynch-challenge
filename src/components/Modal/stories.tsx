import { Story, Meta } from '@storybook/react/types-6-0'

import { useModal } from 'hooks/useModal'
import Modal from '.'

export default {
  title: 'Modal',
  component: Modal
} as Meta

export const Default: Story = () => {
  const { open, setOpen } = useModal()
  return (
    <>
      <button onClick={() => setOpen(!open)}>
        <span>{open ? 'Close Modal' : 'Open Modal'}</span>
      </button>

      <Modal>
        <h1>Modal Content</h1>
      </Modal>
    </>
  )
}
