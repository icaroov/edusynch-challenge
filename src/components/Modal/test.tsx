import { fireEvent, screen } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'

import { renderWithTheme } from 'utils/tests/helpers'
import { useModal } from 'hooks/useModal'

import Modal from '.'

describe('<Modal />', () => {
  it('should render modal when button clicked', () => {
    const { result } = renderHook(() => useModal())

    const { container } = renderWithTheme(
      <>
        <Modal>
          <h1>Modal</h1>
        </Modal>
        <button onClick={result.current.handleOpenModal}>Show Modal</button>
      </>
    )

    // modal closed
    expect(screen.queryByLabelText(/modal/i)).not.toBeInTheDocument()

    // modal open
    const showModalButton = screen.getByText(/show modal/i)

    fireEvent.click(showModalButton)
    expect(screen.getByText(/modal/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
