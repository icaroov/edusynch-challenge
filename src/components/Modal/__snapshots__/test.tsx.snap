// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render modal when button clicked 1`] = `
<button>
  Show Modal
</button>
`;
